lib/virtual.py

```diff
@@ -0,0 +1,144 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2024 IBM
+# Author: Shaik Abdulla <abdulla1@linux.vnet.ibm.com>
+
+"""
+Module for all Virtual devices related functions.
+"""
+
+import re
+import os
+import platform
+
+from .helper import runcmd
+
+
+def get_mac_address(interface):
+    '''
+    Gets the mac_address of given interface.
+    parameter: interface: Name of Interface.
+
+    :return: string of MAC address.
+    '''
+    try:
+        for line in runcmd("ip a s dev %s" % interface)[1].splitlines():
+            if 'link/ether' in line:
+                mac_address = line.split()[1]
+                return mac_address
+    except Exception as e:
+        print(f'Interface not found {e}')
+
+
+def get_driver(interface):
+    '''
+    Gets associated driver/module of given interface.
+    parameter: interface: Name of Interface.
+
+    :return: string of driver name.
+    '''
+    for line in runcmd("ethtool -i  %s" % interface)[1].splitlines():
+        if line.startswith('driver:'):
+            driver = line.split(': ')[1].strip()
+            return driver
+
+def get_veth_interface_names():
+    '''
+    Gets all vETH interface names.
+
+    :return: list of virtual interface names.
+    '''
+    veth_list = []
+    for input_string in runcmd("lsdevinfo -c")[1].splitlines():
+        if 'l-lan' in input_string:
+            pattern = r'name="([^"]+)"'
+            match = re.search(pattern, input_string)
+            if match:
+                # Exclude net0 as its public interface.
+                if not match.group(1) == "net0":
```

Comment:
is `net0`  always an interface in the system or do we have an assumption using our use case? I feel let's have a generic way @abdhaleegit, your view, please

Reply:
@PraveenPenguin usually since "net0" or public interface also type of veth, we wanted to exclude the public interface to use it.
if none of the "veth" interfaces provided by user, then it will try to add up first available veth interface with excluding "net0".